Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp

```diff
@@ -1891,6 +1892,16 @@ void NetworkProcessProxy::deleteWebsiteDataInWebProcessesForOrigin(OptionSet<Web
         if (process.canSendMessage() && !process.isDummyProcessProxy())
             process.sendWithAsyncReply(Messages::WebProcess::DeleteWebsiteDataForOrigin(dataTypes, origin), [callbackAggregator] { });
     }
+    bool shouldClearNavigationSnapshots = dataTypes.contains(WebsiteDataType::MemoryCache) && origin.topOrigin == origin.clientOrigin;
+    if (shouldClearNavigationSnapshots) {
+#if PLATFORM(COCOA) || PLATFORM(GTK)
+        ViewSnapshotStore::singleton().discardSnapshotImagesForOrigin(origin.topOrigin);
+#endif
+        // Since this navigation requested that we clear existing navigation snapshots, we shouldn't
+        // create a snapshot for this navigation either.
+        if (auto page = WebProcessProxy::webPage(webPageProxyID))
+            page->suppressNextAutomaticNavigationSnapshot();
```

Comment:
This surprised me a little bit. The rest of the current navigation is still present in the memory cache, right?

Reply:
I don't understand the question.

Let me give an example which hopefully will clarify:
1. Imagine the Page P is getting navigated same-origin from A1 to A2 and A2 serves the `Clear-Site-Data: "cache"` header
2. When we get the resource response for A2, we notice that the Clear-Site-Data is present is we clear all caches for origin A. Note that at this point, we haven't committed the load to A2 yet so the page is still displaying A1. The specification says we have to clear data before we start rendering the response to the user. This means that at this point, we haven't yet created or back/forward cache entry for A1 and we haven't yet snapshotted A1.
3. Once data is cleared, we eventually commit the load once we start getting the response data.

This means that if `Clear-Site-Data: "cache"` didn't suppress the navigation snapshot, a snapshot of A1 would be taken right after A2 asked us to clear caches for origin A. I had to do something similar for the back/forward cache in an earlier patch. It makes sense for snapshot and back/forward cache since they:
1. occur on navigation, after we process the response, when we commit the load
2. apply to the page we navigated AWAY from

Your comment says "the current navigation is still in the memory cache" but navigations are not stored in the memory cache? Also, the memory cache gets cleared before we commit the load of A2.

Reply:
I don't understand the question.

Let me give an example which hopefully will clarify:
1. Imagine the Page P is getting navigated same-origin from A1 to A2 and A2 serves the `Clear-Site-Data: "cache"` header
2. When we get the resource response for A2, we notice that the Clear-Site-Data is present is we clear all caches for origin A. Note that at this point, we haven't committed the load to A2 yet so the page is still displaying A1. The specification says we have to clear data before we start rendering the response to the user. This means that at this point, we haven't yet created or back/forward cache entry for A1 and we haven't yet snapshotted A1.
3. Once data is cleared, we eventually commit the load once we start getting the response data.

This means that if `Clear-Site-Data: "cache"` didn't suppress the navigation snapshot, a snapshot of A1 would be taken right after A2 asked us to clear caches for origin A. I had to do something similar for the back/forward cache in an earlier patch. It makes sense for snapshot and back/forward cache since they:
1. occur on navigation, after we process the response, when we commit the load
2. apply to the page we navigated AWAY from

Your comment says "the current navigation is still in the memory cache" but navigations are not stored in the memory cache? Also, the memory cache gets cleared before we commit the load of A2.